source/mqtt_client_connection.c

```diff
@@ -345,40 +356,37 @@ PyObject *aws_py_mqtt_client_connection_new(PyObject *self, PyObject *args) {
                 NULL /*validator userdata*/)) {
 
             PyErr_SetAwsLastError();
-            goto use_websockets_failed;
+            goto creation_failed;
         }
     }
 
-    PyObject *self_proxy = PyWeakref_NewProxy(self_py, NULL);
+    self_proxy = PyWeakref_NewProxy(self_py, NULL);
     if (!self_proxy) {
-        goto proxy_new_failed;
+        goto creation_failed;
     }
 
     PyObject *capsule =
         PyCapsule_New(py_connection, s_capsule_name_mqtt_client_connection, s_mqtt_python_connection_destructor);
     if (!capsule) {
-        goto capsule_new_failed;
+        goto creation_failed;
     }
 
     /* From hereon, nothing will fail */
-
-    py_connection->self_capsule = capsule;
     py_connection->self_proxy = self_proxy;
 
     py_connection->client = client_py;
     Py_INCREF(py_connection->client);
 
     return capsule;
 
-capsule_new_failed:
-    Py_DECREF(self_proxy);
-proxy_new_failed:
-use_websockets_failed:
-set_interruption_failed:
-set_connection_handlers_failed:
-    aws_mqtt_client_connection_release(py_connection->native);
-connection_new_failed:
-    aws_mem_release(allocator, py_connection);
+creation_failed:
+    Py_XDECREF(self_proxy);
+    if (py_connection) {
+        if (py_connection->native) {
```

Comment:
You also don't need to null check this either.  release handles a null invocation.  The original suggestion would work verbatim.